File: Algorithm/Python/permutations.py
class Solution:
    def permute(self,num):
        n = len(num)
        if n == 1:
            return [num]
        res=[]
        for i in range(n):
            for subperms in self.permute(num[:i]+num[i+1:]):
                res.append([num[i]]+subperms)
        return res
    def perm(self,arr, pos = 0):
        if pos == len(arr):
            yield arr
            
        for i in range(pos, len(arr)):
            arr[pos], arr[i] = arr[i], arr[pos]#swap the elements sequential
            for x in self.perm(arr, pos + 1): 
                yield x
            arr[pos], arr[i] = arr[i], arr[pos]

if __name__=="__main__":
    sol=Solution()
    num=[1,2,3]
    print sol.permute(num)
    for x in sol.perm(num, 0):
        print x

File: Algorithm/CPP/permutations.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;

class Solution {
public:
	vector<vector<int>> permute(vector<int>& nums)
	{
		vector<vector<int>> res;
		dfs(res, 0,1, nums);
		return res;
	}

	void dfs(vector<vector<int>>& res, int start, int depth,vector<int> nums)
	{
		if (depth == nums.size())
		{
			res.push_back(nums);
			return;
		}
		for (auto i = start; i != nums.size(); ++i)
		{
			swap(nums[i],nums[start]);
			dfs(res,start+1,depth+1,nums );
			swap(nums[i], nums[start]);
		}

	}
};
int  main()
{
	Solution sol;
	vector<int> nums1;
	for (int i = 0; i < 1; ++i)
	{
		nums1.push_back(rand() % 100);

	}
	sort(nums1.begin(), nums1.end());

	for (auto num1 : nums1) cout << num1 << " ";
	cout << endl;
	cout << " the answer is :" << endl;
	for (auto res : sol.permute(nums1))
	{
		for (auto num : res) cout << num << " ";
		cout << endl;
	}
	cout<< endl;
	cin.get();
	return 0;
}
